src/batchproof_container.cpp

```diff
@@ -69,60 +69,50 @@ void BatchProofContainer::add(lelantus::JoinSplit* joinSplit,
-            std::pair<std::pair<uint32_t, bool>, bool> key1 = std::make_pair(std::make_pair(id, false), true);
-            std::pair<std::pair<uint32_t, bool>, bool> key2 = std::make_pair(std::make_pair(id, true), true);
-            std::vector<LelantusSigmaProofData>* vProofs;
-            if (lelantusSigmaProofs.count(key1) > 0)
-                vProofs = &lelantusSigmaProofs[key1];
-            else if (lelantusSigmaProofs.count(key2) > 0)
-                vProofs = &lelantusSigmaProofs[key2];
-            else
-                continue;
-            for (auto dataItr = vProofs->begin(); dataItr != vProofs->end(); dataItr++) {
-                if (dataItr->serialNumber == spendSerial.first) {
-                    vProofs->erase(dataItr);
-                    break;
-                }
-            }
-        }
     }
 }
 void BatchProofContainer::removeLelantus(std::unordered_map<Scalar, int> spentSerials) {
     for (auto& spendSerial : spentSerials) {
+
+        int id = spendSerial.second;
+        int coinGroupId = id % (CENT / 1000);
+        int64_t intDenom = (id - coinGroupId);
+        intDenom *= 1000;
+        sigma::CoinDenomination denomination;
+        bool isSigmaToLela = false;
+        if (sigma::IntegerToDenomination(intDenom, denomination))
+            isSigmaToLela = true;
+
         // afterFixes bool with the pair of set id is considered separate set identifiers, so try to find in one set, if not found try also in another
-        std::pair<std::pair<uint32_t, bool>, bool> key1 = std::make_pair(std::make_pair(spendSerial.second, false), true);
-        std::pair<std::pair<uint32_t, bool>, bool> key2 = std::make_pair(std::make_pair(spendSerial.second, true), true);
+        std::pair<std::pair<uint32_t, bool>, bool> key1 = std::make_pair(std::make_pair(id, false), isSigmaToLela);
+        std::pair<std::pair<uint32_t, bool>, bool> key2 = std::make_pair(std::make_pair(id, true), isSigmaToLela);
         std::vector<LelantusSigmaProofData>* vProofs;
-        if (lelantusSigmaProofs.count(key1) > 0)
+        if (lelantusSigmaProofs.count(key1) > 0) {
             vProofs = &lelantusSigmaProofs[key1];
-        else if (lelantusSigmaProofs.count(key2) > 0)
+            erase(vProofs, spendSerial.first);
+        }
+
+        if (lelantusSigmaProofs.count(key2) > 0) {
             vProofs = &lelantusSigmaProofs[key2];
-        else
-            continue;
+            erase(vProofs, spendSerial.first);
+        }
+    }
+}
 
-        for (auto dataItr = vProofs->begin(); dataItr != vProofs->end(); dataItr++) {
-            if (dataItr->serialNumber == spendSerial.first) {
-                vProofs->erase(dataItr);
-                break;
-            }
+void BatchProofContainer::erase(std::vector<LelantusSigmaProofData>* vProofs, const Scalar& serial) {
+    for (auto dataItr = vProofs->begin(); dataItr != vProofs->end(); dataItr++) {
+        if (dataItr->serialNumber == serial) {
+            vProofs->erase(dataItr);
```

Comment:
This will crash if serial is the last element of `vProofs` (`dataItr` will be past the end of vector). Also `std::vector<T>::erase` will invalidate iterators in some cases. I'd do it with `std::remove_if`.

Reply:
agree, done